tycho-indexer/src/extractor/reorg_buffer.rs

```diff
@@ -105,7 +106,12 @@ where
         &mut self,
         final_block_height: u64,
     ) -> Result<Vec<B>, StorageError> {
-        let target_index = self.find_index(|b| b.block().number == final_block_height);
+        let target_index = if self.last_finalized_block.replace(final_block_height) != Some(final_block_height) {
```

Comment:
So this breaks down to:
```rust
if let Some(prev_final_block_height) = self.last_finalized_block {
   if prev_final_block_height == final_block_height {
       self.find_index(|b| b.block().number == final_block_height).map(|idx| idx + 1)
   } else {
       // Inner case: here may have mixed up buffers, should we throw an error?
       Some(0)
   }
} else {
    // Outer case: this case happens if finalized block height is not yet set
    Some(0)
}
```

So if we have no previously recorded final block height we simply drain the entire buffer and we do the same in case there is a mismatch in finalized block heights. The outer case may be handled by requiring last finalized block height upon initialisation instead of setting it to `None` but I am not too concerned about this one. 

I wonder if in the inner case we shouldn't throw an error instead to be more on the cautious side.

Reply:
Thanks for the breakdown, but there are a few issues with it.

Currently, `split_off(idx)` excludes the block matching `final_block_height`. For example, if the buffer is [20, 21, 22, 23] and final_block_height = 23, `find_index` returns idx = 3, but `split_off(3)` removes 23, leaving [20, 21, 22]. This causes 23 to remain in the buffer.

Regarding the "Inner case" you mentioned, I guess it refers to when `prev_final_block_height == final_block_height`, meaning the finalized_block is unchanged. In this case, idx = 0 applies, and no return blocks is needed. The buffer only needs to be drained when `prev_final_block_height != final_block_height`.

If we expand the details, it should be equivalent to:
```rust
let target_index = if let Some(prev_final_block_height) = self.last_finalized_block {
    if prev_final_block_height == final_block_height {
        Some(0);
    } else {
        self.find_index(|b| b.block().number > final_block_height);
    }
} else{
    Some(0);
};
self.last_finalized_block = Some(final_block_height);
```